### PR TITLE
Add StatefulSet support in OpenTelemetry Operator

### PR DESCRIPTION
**Pull request by IrisTuntun**
Add StatefulSet support in Operator so that the Operator can start collectors in StatefulSet mode.  Also added unit tests and e2e tests. Should resolve  [#31](https://github.com/open-telemetry/wg-prometheus/issues/31) and [#47](https://github.com/open-telemetry/wg-prometheus/issues/47).

### CODE REVIEW

File: config/webhook/manifests.yaml [hunk 1]
```diff
@@ -24,10 +26,11 @@ webhooks:
     - UPDATE
     resources:
     - opentelemetrycollectors
-  admissionReviewVersions: ["v1","v1beta1"]
   sideEffects: None
-- clientConfig:
-    caBundle: Cg==
```
Comment (Aneurysm9) [inline config/webhook/manifests.yaml:30]: It looks like, other than rearranging keys, the only effect here is to remove this default `caBundle` value.  Why is this done?

File: pkg/collector/container.go [hunk 2]
```diff
@@ -56,6 +56,11 @@ func Container(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelem
 
 	if len(otelcol.Spec.VolumeMounts) > 0 {
 		volumeMounts = append(volumeMounts, otelcol.Spec.VolumeMounts...)
+	} else if otelcol.Spec.Mode == "statefulset" {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      "initial-volume",
```
Comment (Aneurysm9) [inline pkg/collector/container.go]: Maybe `default-volume` would be a better name here?

What happens if a user provides non-default VolumeClaimTemplate entries but doesn't specify any VolumeMounts?

File: pkg/collector/statefulset.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/open-telemetry/opentelemetry-operator/api/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/internal/config"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/naming"
+)
+
+// StatefulSet builds the deployment for the given instance.
```
Comment (Aneurysm9) [inline pkg/collector/statefulset.go]:
```suggestion
// StatefulSet builds the statefulset for the given instance.
```

**Review by Aneurysm9 — APPROVED** [review]
LGTM, I think the questions I've asked here can be carried over and addressed in the upstream PR.

File: config/webhook/manifests.yaml (hunk 1 above)
Comment (humivo) [reply config/webhook/manifests.yaml:30]: I'm not exactly sure why that was done but I think it due to the changes we made to the Makefile where we change the controller-tools version and etc.  It does not seem to have any outstanding effects on the Operator though.

File: pkg/collector/container.go (hunk 2 above)
Comment (humivo) [reply pkg/collector/container.go]: If a user provides a non-default VolumeClaimTemplate entry and no corresponding VolumeMount, then the Operator will fail to create the StatefulSet as it is required.

**Review by alolita — APPROVED** [review]
+1

**Pull request closed without merging**